src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -76,64 +78,129 @@ data IgnoreMode
-        (File, _)              -> throwError (makeError mode "" pos)
-        (Link, LINK {})        -> put None $> defNode
-        (Link, _)              -> Node pos ty <$> sequence subs
-        (None, _) -> do
-          case getIgnoreMode (Node pos ty []) of
-            Just mode' -> put mode' $> defNode
+        -- We expect to find a paragraph immediately after the
+        -- `ignore paragraph` annotanion. If the paragraph is not
+        -- found we should report an error.
+        (Paragraph, PARAGRAPH) -> put ignoreNone $> defNode
+        (Paragraph, x)         -> do
+          lift . tell . makeError pos' fp mode $ prettyType x
+          put ignoreNone
+          Node pos ty <$> sequence subs
+
+        -- We don't expect to find an `ignore file` annotation here,
+        -- since that annotation should be at the top of the file and
+        -- the file should already be ignored when `checkIgnoreFile` is called.
+        -- We should report an error if we find it anyway.
+        (File, _)              -> do
+          lift . tell $ makeError pos' fp mode ""
+          put ignoreNone
+          Node pos ty <$> sequence subs
+
+        -- When we find an `ignore link annotation`, we skip nodes until
+        -- we find a link and ignore it, or we find another ignore annotation,
+        -- then we should report an error and set new `Ignore` state.
+        (Link, LINK {})        -> put ignoreNone $> defNode
+        (Link, _)            ->
+          case getIgnoreMode node of
+            Just mode'  -> do
+              lift . tell $ makeError pos' fp mode ""
+              handleMode node mode'
+            Nothing     -> Node pos ty <$> sequence subs
+
+        -- When no `Ignore` state is set check next node for annotation,
+        -- if found then set it as new `IgnoreMode` otherwise skip node.
+        (None, _)              ->
+          case getIgnoreMode node of
+            Just mode' -> handleMode node mode'
             Nothing    -> Node pos ty <$> sequence subs
 
+    handleMode
+      :: Node
+      -> IgnoreMode
+      -> ScannerM Node
+    handleMode node = \case
+      -- Report unknown `IgnoreMode`.
+      None   -> do
+        let unrecognised = fromMaybe ""
+              $ safeHead . drop 1 . words =<< getXrefcheckContent node
+        lift . tell $ makeError (getPosition node) fp None unrecognised
+        put ignoreNone $> defNode
+      -- Set new `Ignore` state.
+      mode'  -> put (Ignore mode' $ getPosition node) $> defNode
+
     prettyType :: NodeType -> Text
     prettyType ty =
       let mType = safeHead $ words $ show ty
       in fromMaybe "" mType
 
+    withIgnoreMode
+      :: ScannerM Node
+      -> Writer [ScanError] Node
+    withIgnoreMode action = action `runStateT` ignoreNone >>= \case
+      -- We expect `IgnoreMode` to be `None` when we reach EOF,
+      -- otherwise that means there was an annotation that didn't match
+      -- any node, so we have to report that.
```

Comment:
We should probably test this scenario (`ignore link` and `ignore paragraph` at the end of the file)